### ISSUE TITLE
Single version support multilevel select API

### ISSUE THREAD

**Issue by tenmozes** [VictoriaMetrics/VictoriaMetrics#4328]
### Is your feature request related to a problem? Please describe

Need to backport functionality from vmslect to support multi-level setup 

```
vmselect nodes can be queried by other vmselect nodes if they run with -clusternativeListenAddr command-line flag. For example, if vmselect is started with -clusternativeListenAddr=:8401, then it can accept queries from another vmselect nodes at TCP port 8401 in the same way as vmstorage nodes do. This allows chaining vmselect nodes and building multi-level cluster topologies. For example, the top-level vmselect node can query second-level vmselect nodes in different availability zones (AZ), while the second-level vmselect nodes can query vmstorage nodes in local AZ.
```

### Describe the solution you'd like

vmselect can query single node
tenant id should be a part of the query parameter or flag parameter

**Comment by zoglam** [comment]
Any updates?

**Comment by proffust** [comment]
Adding compatibility between single-node and cluster deployments would enable quick and seamless migration between these versions. For example, users could start with vmsingle, and when resources become insufficient, they could add nodes and configure vmselect to read data from the single node while writing new metrics to the cluster. This would provide a smooth transition path from single-node to cluster setup without data loss or service interruption.

Currently, migration between vmsingle and vmcluster is a time-consuming operation, which creates significant operational overhead and inconvenience.

This feature would be particularly useful for:
- Growing deployments that need to scale
- Testing cluster setup while keeping single-node as a fallback
- Hybrid setups during transition periods

**Comment by Kot-o-pes** [comment]
This is also can be useful for non-homogeneous deployments to create single plane of glass for both, clusters and single installations.

**Comment by jiekun** [comment]
I think the different data structure is one of the blocker of this feature. @Kot-o-pes @proffust

**Comment by valyala** [comment]
The main reason for the lack of vmstorage API at VictoriaMetrics single-node is that we haven't decided how to handle requests from vmselect for different tenants. Single-node VictoriaMetrics doesn't support tenants, and vmselect sends requests to the vmstorage API specifying a tenant. We considered three options:
1. To ignore the tenant in the request from vmselect, and to return the same result for any requested tenant.
2. To execute the request only for tenant id 0:0, and return an empty response for the rest of the tenants.
3. To provide the ability to specify the tenant id at single-node VictoriaMetrics via a command-line flag for which to execute the request, and to return an empty response for the rest of the requested tenants. The default tenant to request must be 0:0. The command&line flag name could be `-clusternativeTenantID` in order to be consistent with the `-clusternativeListenAddr` (see [these docs](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#multi-level-cluster-setup)).

After we decide which option to follow, the implementation of the vmstorage API on a single-node VictoriaMetrics shouldn't be too hard.

Personally I like the third option, since it works like the second option when the tenant id isn't specified at single-node VictoriaMetrics (this should cover the majority of practical use cases), while it allows specifying other tenants on a per-single-node basis if needed.

@makasim , let's prioritise and implement this feature, since it nicely complements [the existing multi-level cluster setup for querying API](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#multi-level-cluster-setup), and it doesn't interfere with the existing code too much. E.g. it can be considered as an improvement of the existing feature.

**Comment by makasim** [comment]
> Personally I like the third option, since it works like the second option when the tenant id isn't specified at single-node VictoriaMetrics (this should cover the majority of practical use cases), while it allows specifying other tenants on a per-single-node basis if needed.

@f41gh7 and I also like p3 option.

**Comment by valyala** [comment]
Let's prioritize this feature because it solves the following important cases:

- It simplifies the migration path from single-node to cluster version of VictoriaMetrics. When you reach single-node vertical scalability limits, you can convert this single-node into cluster by running vmselect in front of it. The migration is instant - you don't need to migrate the data from single-node to cluster. You can scale horizontally the capacity and the performance of the cluster by adding regular vmstorage nodes to it. vminsert must be configured to write data only to vmstorage nodes, and not to the single-node VictoriaMetrics. E.g. the single-node VictoriaMetrics must be in read-only mode. Later you can decommission the single-node instance from the cluster after all the data is automatically removed from it because of the configured retention.
- It allows replacing promxy with vmselect and get access to all the MetricsQL features. See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4551 .
- It allows building flexible global querying view schemes over the mix of multiple single-node instances, VictoriaMetrics clusters and vmstorage nodes.

**Comment by makasim** [comment]
@valyala  It’s on my list. I haven’t pushed it forward yet because I’m still evaluating other options. While this feature would benefit clients, it doesn’t reduce the operational overhead of maintaining four different VictoriaMetrics variants. My goal is to find a solution that addresses both. For example, merging single-node and cluster versions could deliver the same client benefits while significantly reducing operational complexity. I’m exploring ways to achieve this while preserving backward compatibility and avoiding excessive complexity.

**Comment by ashutoshdcx** [comment]
@makasim Eagerly waiting to hear back some updates on this. 
TIA.

**Comment by makasim** [comment]
@ashutoshdcx We are working on it https://github.com/VictoriaMetrics/VictoriaMetrics/pull/9578. Eventually, the feature will be released, cannot say when this will happen.

**Comment by zoglam** [comment]
Hi, @makasim 
I noticed that the MR #9578 was closed without being merged. Could you please share what happened?
TIA.

**Comment by makasim** [comment]
We are going to implement this feature.  It is on our short list. ~~I reopened PR~~ Will be implemented in another PR

**Comment by zoglam** [comment]
Sorry for the ping!😶‍🌫️ @makasim, any updates on this? An ETA would be really helpful or link to PR. Thanks!

**Comment by zoglam** [comment]
Hi @makasim, sorry to ping again. Just checking if there are any updates since the last comment? Appreciate your help

**Comment by makasim** [comment]
I'll provide an update once there is any. There is no need to ping me.

**Comment by ben-dov** [comment]
Hey,
Its nice to see that this is being worked on.

A question about the implementation, will it also allow us to use vmselect as `-remoteRead.url` in vmalert?
Specifically needed in a two vmsingle HA scenario, since promxy does not work.

**Comment by f41gh7** [comment]
> Hey, Its nice to see that this is being worked on.
> 
> A question about the implementation, will it also allow us to use vmselect as `-remoteRead.url` in vmalert? Specifically needed in a two vmsingle HA scenario, since promxy does not work.

 Yes, it will be possible to use it in such scenario, but you have to deploy vmselect as an entry point for requests.

 Configuration will be the following:
1) vmselect 
vmselect -storageNode=vmsingle-0:8401,vmsingle-1:8401 -dedupMinScrapeInterval=1ms
2) vmsingle-0 with `vmselectAddr=:8401`
3) vmsingle-1 with `vmselectAddr=:8401`

**Comment by makasim** [comment]
In https://github.com/VictoriaMetrics/VictoriaMetrics/pull/10926, `vmsingle` gained support for the cluster-native read API. This allows it to serve as a `vmstorage` backend in a [VictoriaMetrics cluster ](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/) for read (select) requests only.

The primary use case is simplifying migration from a single-node deployment to a cluster. Instead of migrating historical data, you can:
* Deploy a new VictoriaMetrics cluster and start ingesting new data into it.
* Add the existing `vmsingle` instance as a read-only `vmstorage` backend so historical data remains accessible.
* Once the retention period expires on `vmsingle`, decommission it.

This enables a seamless migration without data copying or downtime.

**Comment by makasim** [comment]
The feature included in [v1.147.0](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/victoriametrics/changelog/CHANGELOG.md#v11470) release.

**Comment by makasim** [comment]
Read how to migrate from a single node to cluster in https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#from-single-node-to-cluster.

Read how to build multitenancy on single node instances in https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#multi-tenancy
